README.md
<div align="center">
	<p align="center">
		<img src="https://raw.githubusercontent.com/ali-master/pwa-badge/master/images/image.png" border="0" />
	</p>
	<h1 align="center">PWA Badge</h1>
	<p align="center">Like <b>Native Apps</b>, The PWA Badge API allows installed web apps to set an application-wide badge on the app icon.</p>

[![Travis CI](https://travis-ci.com/ali-master/pwa-badge.svg?branch=master)](https://travis-ci.com/ali-master/pwa-badge)
[![codecov](https://codecov.io/gh/ali-master/pwa-badge/branch/master/graph/badge.svg)](https://codecov.io/gh/ali-master/pwa-badge)
[![CodeFactor](https://www.codefactor.io/repository/github/ali-master/pwa-badge/badge)](https://www.codefactor.io/repository/github/ali-master/pwa-badge)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/ali-master/pwa-badge/blob/master/LICENSE)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-orange.svg)](https://github.com/ali-master/pwa-badge/compare)
[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg)](#contributors-)
</div>
<hr />

The App Badge API allows installed web apps to set an application-wide badge,
shown in an operating-system-specific place associated with the application
(such as the shelf or home screen).

Badges tend to be more user-friendly than notifications and can be updated with
a much higher frequency since they don't interrupt the user. And, because they
don't interrupt the user, they don't need the user's permission.

> Keep in mind that to display the Badge count, Your **PWA** application should be installed on your Device.

### Possible use cases

Examples of sites that may use this Library includes:

- Chat, email, and social apps, to signal that new messages have arrived, or to
  show the number of unread items.
- Productivity apps, to signal that a long-running background task (such as
  rendering an image or video) has completed.
- Games, to signal that a player action is required (e.g., in Chess, when it is
  the player's turn).

### Usage

```bash
npm install --save pwa-badge
```

### The Badge API consists of five methods:

- `isSupported()` Check if the User's browser supports the Feature,
  and returns a `boolean` value that represents the Status of supporting.
- `syncSetBadge(unreadCount)` Removes app's badge **Synchronously**. If a value
  is provided, set the badge to the provided value otherwise, display a plain
  white dot (or other flag as appropriate to the platform). Setting number to 0
  is the same as calling `syncClearBadge()` or `asyncClearBadge()`.
- `syncClearBadge()` Removes app's badge **Synchronously**.
- `asyncSetBadge(unreadCount)` This API is the same as `syncSetBadge()` but
  returns an empty `Promise` for error handling.
- `asyncClearBadge()` Removes app's badge **Asynchronously** and returns an
  empty `Promise` for error handling.

### Check Browser supports the Badge API

TL;DR `isSupported()` method function is an util for informing your users that
this feature supports by their `Browser` or `OS` and the `pwa-badge` library
`set` and `clear` the Badge count safely, and you can avoid using
`isSupported()` before calling the `set` or `clear` methods.

```js
import PWABadge from 'pwa-badge';

// Create an Instance
const badge = new PWABadge();

if (badge.isSupported()) {
  // Hoora!, Supports the Badge feature
} else {
  // Does not supports
}
```

### `Sync` Set and Clear Badge

```js
import PWABadge from 'pwa-badge';

// Create an Instance
const badge = new PWABadge();

// Set Badge unreadCount
badge.syncSetBadge(1);

// Clear Badge unreadCount
badge.syncClearBadge();
```

Result by calling `syncSetBadge`:

<div align="center">
	<p align="center">
		<img src="https://raw.githubusercontent.com/ali-master/pwa-badge/master/images/demo.png" border="0" />
	</p>
</div>

### `Async` Set and Clear Badge

`asyncSetBadge()` and `asyncClearBadge()` return empty `promises` you can
use for error handling.

```js
import PWABadge from 'pwa-badge';

// Create an Instance
const badge = new PWABadge();

// Set Badge unreadCount
badge
  .asyncSetBadge(1)
  .then(() => {
    // Badge count has shown as well
  })
  .catch((e) => {
    // The Browser not supporting the Badge feature or something went wrong
  });

// Clear Badge unreadCount
badge
  .asyncClearBadge()
  .then(() => {
    // Badge count has disappeared
  })
  .catch((e) => {
    // The Browser not supporting the Badge feature or something went wrong
  });
```

The App Badge API works on **Windows**, and **macOS**, in **Chrome 81 or
later**. It has also been confirmed to work on **Edge 84** or later. Support for
**Chrome OS** is in development and will be available in a future release of
Chrome. On **Android**, the Badge API is not supported. Instead, Android
automatically shows a badge on app icon for the installed web app when there is
an unread notification, just as for Android apps.

Some user agents may take a number like `4000` and rewrite it as `99+`. If you
saturate the badge yourself (for example by setting it to `99`) then the `+`
won't appear. No matter the actual number, just call `syncSetBadge()` or
`asyncSetBadge()` and let the user agent deal with displaying it accordingly.

While the App Badge API in Chrome requires an `installed app` as I wrote
before, you shouldn't make calls to the Badge API dependent on the
installation state. Just call the API when it `exists` and `installed` on a
device, as other browsers may show the badge in other places. If it works, it
works. If not, it simply doesn't.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[standard-version](https://github.com/conventional-changelog/standard-version)
for commit guidelines.

### [1.1.1](https://github.com/ali-master/pwa-badge/compare/v1.1.0...v1.1.1) (2021-01-12)

### Bug Fixes

- Missing bundle files
  ([639f79b](https://github.com/ali-master/pwa-badge/commit/639f79b21762fab5a5971783d389d33843007052))

## [1.1.0](https://github.com/ali-master/pwa-badge/compare/v1.0.1...v1.1.0) (2021-01-12)

### Features

- Added readonly to the private keys in index.d.ts and improved jsdoc
  ([e887132](https://github.com/ali-master/pwa-badge/commit/e887132576a195d10f1a656bd21179e064136063))
- Added standard-version for creating changelog and auto versioning
  ([12c1692](https://github.com/ali-master/pwa-badge/commit/12c169234542911e5566f1a0715de3cb936c5bd9))
- Run tests before running release command
  ([7b9c26e](https://github.com/ali-master/pwa-badge/commit/7b9c26ea5c1a1799d875b0de27ebb0b08fccc9ce))

### Bug Fixes

- CI build
  ([eb37614](https://github.com/ali-master/pwa-badge/commit/eb37614657a3846aadef13374cbb6d90b1f6b0bb))
- Eslint and Travis issues
  ([e216860](https://github.com/ali-master/pwa-badge/commit/e216860988e5e379b33cd71dc157f009a1ca32f6))
- syncClearBadge param
  ([8c346f7](https://github.com/ali-master/pwa-badge/commit/8c346f7eac253a98babcf27a1298d25d02352aeb))


test/index.js
// import chai from 'chai';

describe('PWA badge', () => {

});


test/index.spec.js
/* eslint-disable @typescript-eslint/no-empty-function */
import PWABadge from '../src/index.js';

function clearNavigator() {
  Object.defineProperty(window, 'navigator', {
    value: {},
    enumerable: true,
    writable: true,
    configurable: true,
  });
}

function clearWindow() {
  delete window.ExperimentalBadge;
}
let Badge;
describe('PWA Badge', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.resetModules();
    clearNavigator();
    clearWindow();
    Badge = new PWABadge();
  });
  describe('Check isSupported by navigator and window methods', () => {
    it('Should be false', async () => {
      expect(Badge.isSupported()).toEqual(false);
      expect(Badge.syncSetBadge(10)).toBeUndefined();
      await expect(Badge.asyncSetBadge(10)).rejects.toBeUndefined();
      await expect(Badge.asyncClearBadge()).rejects.toBeUndefined();
    });
    it('Should be true By using setBadge', () => {
      Object.defineProperty(navigator, 'setBadge', {
        value: async (unreadCount) => {
          return await unreadCount;
        },
        enumerable: true,
      });
      expect(Badge.isSupported()).toBe(true);
    });

    it('Should be true By using setExperimentalBadge', () => {
      Object.defineProperty(navigator, 'setExperimentalBadge', {
        value: async (unreadCount) => {
          return await unreadCount;
        },
        enumerable: true,
      });
      expect(Badge.isSupported()).toBe(true);
    });

    it('Should be true By using setClientBadge', () => {
      Object.defineProperty(navigator, 'setClientBadge', {
        value: async (unreadCount) => {
          return await unreadCount;
        },
        enumerable: true,
      });
      expect(Badge.isSupported()).toBe(true);
    });

    it('Should be true By using window.ExperimentalBadge', () => {
      Object.defineProperty(window, 'ExperimentalBadge', {
        value: {
          set: async (unreadCount) => {
            return await unreadCount;
          },
        },
        enumerable: true,
      });
      expect(Badge.isSupported()).toBe(true);
    });
  });
  describe('Set and Clear Badge count by navigator.setBadge', () => {
    beforeEach(() => {
      if (!('setBadge' in window.navigator)) {
        Object.defineProperties(window.navigator, {
          setBadge: {
            value: async (unreadCount) => {
              return await unreadCount;
            },
            writable: true,
            enumerable: true,
          },
          clearBadge: {
            value: async () => {
              throw new Error('Something went wrong');
            },
            writable: true,
            enumerable: true,
          },
        });
      }
      jest.spyOn(window.navigator, 'setBadge');
      jest.spyOn(window.navigator, 'clearBadge');
    });

    it('[Sync] Should call navigator.setBadge', async () => {
      expect(Badge.syncSetBadge(10)).toBeUndefined();
      expect(window.navigator.setBadge).toHaveBeenCalledWith(10);
      expect(window.navigator.setBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.setBadge(10)).resolves.toEqual(10);
    });

    it('[Async] Should call navigator.setBadge', async () => {
      await expect(Badge.asyncSetBadge(10)).resolves.toBeDefined();
      expect(window.navigator.setBadge).toHaveBeenCalledWith(10);
      expect(window.navigator.setBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.setBadge(10)).resolves.toEqual(10);
    });

    it('[Sync] Should call navigator.clearBadge', async () => {
      expect(Badge.syncClearBadge()).toBeUndefined();
      expect(window.navigator.clearBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.clearBadge()).rejects.toThrowError(
        'Something went wrong',
      );
    });

    it('[Async] Should call navigator.clearBadge', async () => {
      await expect(Badge.asyncClearBadge()).rejects.toThrowError(
        'Something went wrong',
      );
      expect(window.navigator.clearBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.clearBadge()).rejects.toThrowError(
        'Something went wrong',
      );
    });
  });
  describe('Set and Clear Badge count by navigator.setExperimentalBadge', () => {
    beforeEach(() => {
      if (!('setExperimentalBadge' in window.navigator)) {
        Object.defineProperties(window.navigator, {
          setExperimentalBadge: {
            value: async (unreadCount) => {
              return await unreadCount;
            },
            writable: true,
            enumerable: true,
          },
          clearExperimentalBadge: {
            value: async () => {
              throw new Error('Something went wrong');
            },
            writable: true,
            enumerable: true,
          },
        });
      }
      jest.spyOn(window.navigator, 'setExperimentalBadge');
      jest.spyOn(window.navigator, 'clearExperimentalBadge');
    });

    it('[Sync] Should call navigator.setExperimentalBadge', async () => {
      expect(Badge.syncSetBadge(10)).toBeUndefined();
      expect(window.navigator.setExperimentalBadge).toHaveBeenCalledWith(10);
      expect(window.navigator.setExperimentalBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.setExperimentalBadge(10)).resolves.toEqual(
        10,
      );
    });

    it('[Async] Should call navigator.setExperimentalBadge', async () => {
      await expect(Badge.asyncSetBadge(10)).resolves.toBeDefined();
      expect(window.navigator.setExperimentalBadge).toHaveBeenCalledWith(10);
      expect(window.navigator.setExperimentalBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.setExperimentalBadge(10)).resolves.toEqual(
        10,
      );
    });

    it('[Sync] Should call navigator.clearExperimentalBadge', async () => {
      expect(Badge.syncClearBadge()).toBeUndefined();
      expect(window.navigator.clearExperimentalBadge).toHaveBeenCalledWith();
      expect(window.navigator.clearExperimentalBadge).toHaveBeenCalledTimes(1);
      await expect(
        window.navigator.clearExperimentalBadge(),
      ).rejects.toThrowError('Something went wrong');
    });

    it('[Async] Should call navigator.clearExperimentalBadge', async () => {
      await expect(Badge.asyncClearBadge()).rejects.toThrowError(
        'Something went wrong',
      );
      expect(window.navigator.clearExperimentalBadge).toHaveBeenCalledWith();
      expect(window.navigator.clearExperimentalBadge).toHaveBeenCalledTimes(1);
      await expect(
        window.navigator.clearExperimentalBadge(),
      ).rejects.toThrowError('Something went wrong');
    });
  });
  describe('Set and Clear Badge count by navigator.setClientBadge', () => {
    beforeEach(() => {
      if (!('setClientBadge' in window.navigator)) {
        Object.defineProperties(window.navigator, {
          setClientBadge: {
            value: async (unreadCount) => {
              return await unreadCount;
            },
            writable: true,
            enumerable: true,
          },
          clearClientBadge: {
            value: async () => {
              throw new Error('Something went wrong');
            },
            writable: true,
            enumerable: true,
          },
        });
      }
      jest.spyOn(window.navigator, 'setClientBadge');
      jest.spyOn(window.navigator, 'clearClientBadge');
    });

    it('[Sync] Should call navigator.setClientBadge', async () => {
      expect(Badge.syncSetBadge(10)).toBeUndefined();
      expect(window.navigator.setClientBadge).toHaveBeenCalledWith(10);
      expect(window.navigator.setClientBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.setClientBadge(10)).resolves.toEqual(10);
    });

    it('[Async] Should call navigator.setClientBadge', async () => {
      await expect(Badge.asyncSetBadge(10)).resolves.toBeDefined();
      expect(window.navigator.setClientBadge).toHaveBeenCalledWith(10);
      expect(window.navigator.setClientBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.setClientBadge(10)).resolves.toEqual(10);
    });

    it('[Sync] Should call navigator.clearClientBadge', async () => {
      expect(Badge.syncClearBadge()).toBeUndefined();
      expect(window.navigator.clearClientBadge).toHaveBeenCalledWith();
      expect(window.navigator.clearClientBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.clearClientBadge()).rejects.toThrowError(
        'Something went wrong',
      );
    });

    it('[Async] Should call navigator.clearClientBadge', async () => {
      await expect(Badge.asyncClearBadge()).rejects.toThrowError(
        'Something went wrong',
      );
      expect(window.navigator.clearClientBadge).toHaveBeenCalledWith();
      expect(window.navigator.clearClientBadge).toHaveBeenCalledTimes(1);
      await expect(window.navigator.clearClientBadge()).rejects.toThrowError(
        'Something went wrong',
      );
    });
  });

  describe('Set and Clear Badge count by window.ExperimentalBadge', () => {
    beforeEach(() => {
      if (!('ExperimentalBadge' in window)) {
        Object.defineProperty(window, 'ExperimentalBadge', {
          value: {},
          enumerable: true,
          configurable: true,
          writable: true,
        });
      }
      Object.defineProperty(window.ExperimentalBadge, 'set', {
        value: async (unreadCount) => {
          return await unreadCount;
        },
        enumerable: true,
        writable: true,
      });
      Object.defineProperty(window.ExperimentalBadge, 'clear', {
        value: async () => {
          throw new Error('Something went wrong');
        },
        enumerable: true,
        writable: true,
      });
    });
    it('[Sync] Should call window.ExperimentalBadge.set', () => {
      jest.spyOn(window.ExperimentalBadge, 'set');
      expect(Badge.syncSetBadge(10)).toBeUndefined();
      expect(window.ExperimentalBadge.set).toHaveBeenCalledWith(10);
      expect(window.ExperimentalBadge.set).toHaveBeenCalledTimes(1);
      expect(window.ExperimentalBadge.set(10)).resolves.toEqual(10);
    });

    it('[Async] Should call window.ExperimentalBadge.set', async () => {
      jest.spyOn(window.ExperimentalBadge, 'set');
      await expect(Badge.asyncSetBadge(10)).resolves.toBeDefined();
      expect(window.ExperimentalBadge.set).toHaveBeenCalledWith(10);
      expect(window.ExperimentalBadge.set).toHaveBeenCalledTimes(1);
      await expect(window.ExperimentalBadge.set(10)).resolves.toEqual(10);
    });

    it('[Sync] Should call window.ExperimentalBadge.clear', async () => {
      jest.spyOn(window.ExperimentalBadge, 'clear');
      expect(Badge.syncClearBadge()).toBeUndefined();
      expect(window.ExperimentalBadge.clear).toHaveBeenCalledWith();
      expect(window.ExperimentalBadge.clear).toHaveBeenCalledTimes(1);
      await expect(window.ExperimentalBadge.clear()).rejects.toThrowError(
        'Something went wrong',
      );
    });

    it('[Async] Should call window.ExperimentalBadge.clear', async () => {
      jest.spyOn(window.ExperimentalBadge, 'clear');
      await expect(Badge.asyncClearBadge()).rejects.toThrowError(
        'Something went wrong',
      );
      expect(window.ExperimentalBadge.clear).toHaveBeenCalledWith();
      expect(window.ExperimentalBadge.clear).toHaveBeenCalledTimes(1);
      await expect(window.ExperimentalBadge.clear()).rejects.toThrowError(
        'Something went wrong',
      );
    });
  });
});


test/setupTests.js
global.console = {
  log: jest.fn(),
  error: jest.fn(),
};


src/index.d.ts
declare class PWABadge {
  private readonly navigator: Navigator;
  private readonly window: Window;

  /**
   * Check the Browser Badge feature supports
   */
  isSupported(): boolean;

  /**
   * Sets the PWA App's badge.
   *
   * If a value is provided, set the badge to the provided value otherwise, display a plain white dot (or other flag as reprobate to the platform)
   * Don't assume anything about how the user agent displays the badge.
   * Some user agents may take a number like "4000" and rewrite it as "99+".
   * If you saturate the badge yourself (for example by setting it to "99") then the "+" won't appear.
   * No matter the actual number, just call setAppBadge(unreadCount) and let the user agent deal with displaying it accordingly.
   *
   * Setting number to `0` is the same as calling {@link syncClearBadge|this.syncClearBadge()} or {@link syncClearBadge|this.asyncClearBadge()}.
   *
   * @template Unread Badge count
   */
  syncSetBadge(unreadCount: number): void;

  /**
   * Sets the PWA App's badge.
   *
   * If a value is provided, set the badge to the provided value otherwise, display a plain white dot (or other flag as reprobate to the platform)
   * Don't assume anything about how the user agent displays the badge.
   * Some user agents may take a number like "4000" and rewrite it as "99+".
   * If you saturate the badge yourself (for example by setting it to "99") then the "+" won't appear.
   * No matter the actual number, just call setAppBadge(unreadCount) and let the user agent deal with displaying it accordingly.
   *
   * Setting number to `0` is the same as calling {@link syncClearBadge|this.syncClearBadge()} or {@link syncClearBadge|this.asyncClearBadge()}.
   *
   * @template Unread Badge count
   */
  asyncSetBadge(unreadCount: number): Promise<void>;

  /**
   * Removes app's badge.
   */
  syncClearBadge(): void;

  /**
   * Removes app's badge.
   */
  asyncClearBadge(): Promise<void>;
}

export default PWABadge;


src/index.js
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

/**
 * Badging for PWA app icons
 *
 * The App Badging API allows installed web apps to set an application-wide badge on the app icon.
 *
 * In some cases, the OS may not allow the exact representation of the badge.
 * In such cases, the browser will attempt to provide the best representation for that device.
 * For example, because the Badging API isn't supported on Android, Android only ever shows a dot instead of a numeric value.
 *
 * @class
 * @namespace PWABadge
 * @example Sync Set Badge.
 * const AppBadge = new PWABadge();
 * AppBadge.syncSetBadge(10)
 * @example Async Set Badge
 * const AppBadge = new PWABadge();
 * AppBadge.asyncSetBadge(10).then(() => {
 *   // Do something...
 * }).catch(e => {
 *   // Do something with the error.
 * })
 * @example Sync Clear Badge value
 * const AppBadge = new PWABadge();
 * AppBadge.syncClearBadge()
 * @example Async Clear Badge value
 * AppBadge.asyncClearBadge().then(() => {
 *   // Do something...
 * }).catch(e => {
 *   // Do something with the error.
 * })
 */
class PWABadge {
  constructor() {
    /**
     * @private
     * @type        {Navigator}
     * @description A Clone of navigator object
     */
    this.__navigator = window.navigator;
    /**
     * @private
     * @type        {Window}
     * @description A Clone of window object
     */
    this.__window = window;
  }

  /**
   * Check the Browser Badge feature supports
   *
   * @memberof  PWABadge
   * @return    {boolean} Return true if the browser supports the Badge feature and returns false if not.
   */
  isSupported() {
    const n = ['setBadge', 'setExperimentalBadge', 'setClientBadge'];

    for (const key of n) {
      if (this.__navigator.hasOwnProperty(key)) return true;
    }

    return this.__window.hasOwnProperty('ExperimentalBadge');
  }

  /**
   * Wrapper to support first and second origin trial to compatible with the Browsers
   *
   * @private
   * @memberof  PWABadge
   * @see       https://web.dev/badging-api/
   * @see       https://github.com/w3c/badging/blob/master/docs/implementation.md
   * @returns   {Promise<void>} returns empty promise you can use for error handling
   */
  __innerSet(unreadCount) {
    if (this.__navigator.setBadge) {
      return this.__navigator.setBadge(unreadCount);
    } else if (this.__navigator.setExperimentalBadge) {
      return this.__navigator.setExperimentalBadge(unreadCount);
    } else if (this.__navigator.setClientBadge) {
      return this.__navigator.setClientBadge(unreadCount);
    } else if (this.__window.ExperimentalBadge) {
      return this.__window.ExperimentalBadge.set(unreadCount);
    }

    return Promise.reject();
  }

  /**
   * Sets the PWA App's badge.
   *
   * If a value is provided, set the badge to the provided value otherwise, display a plain white dot (or other flag as reprobate to the platform)
   * Don't assume anything about how the user agent displays the badge.
   * Some user agents may take a number like "4000" and rewrite it as "99+".
   * If you saturate the badge yourself (for example by setting it to "99") then the "+" won't appear.
   * No matter the actual number, just call setAppBadge(unreadCount) and let the user agent deal with displaying it accordingly.
   *
   * Setting number to `0` is the same as calling {@link syncClearBadge|this.syncClearBadge()} or {@link syncClearBadge|this.asyncClearBadge()}.
   *
   * @memberof    PWABadge
   * @param       {number} unreadCount - Unread Badge count
   * @returns     {void}   returns empty promise you can use for error handling
   */
  syncSetBadge(unreadCount) {
    try {
      this.__innerSet(unreadCount);
    } catch (e) {}
  }

  /**
   * Sets the PWA App's badge.
   *
   * If a value is provided, set the badge to the provided value otherwise, display a plain white dot (or other flag as reprobate to the platform)
   * Don't assume anything about how the user agent displays the badge.
   * Some user agents may take a number like "4000" and rewrite it as "99+".
   * If you saturate the badge yourself (for example by setting it to "99") then the "+" won't appear.
   * No matter the actual number, just call setAppBadge(unreadCount) and let the user agent deal with displaying it accordingly.
   *
   * Setting number to `0` is the same as calling {@link syncClearBadge|this.syncClearBadge()} or {@link syncClearBadge|this.asyncClearBadge()}.
   *
   * @async
   * @memberof    PWABadge
   * @param       {number} unreadCount - Unread Badge count
   * @returns     {Promise<void>} returns empty promise you can use for error handling
   */
  async asyncSetBadge(count) {
    return this.__innerSet(count);
  }

  /**
   * Wrapper to support first and second origin trial to compatible with the Browsers
   *
   * @private
   * @memberof PWABadge
   * @see      https://web.dev/badging-api/
   * @see      https://github.com/w3c/badging/blob/master/docs/implementation.md
   * @returns  {Promise<void>} returns empty promise you can use for error handling
   */
  __innerClear() {
    if (this.__navigator.clearBadge) {
      return this.__navigator.clearBadge();
    } else if (this.__navigator.clearExperimentalBadge) {
      return this.__navigator.clearExperimentalBadge();
    } else if (this.__navigator.clearClientBadge) {
      return this.__navigator.clearClientBadge();
    } else if (this.__window.ExperimentalBadge) {
      return this.__window.ExperimentalBadge.clear();
    }

    return Promise.reject();
  }

  /**
   * Removes app's badge.
   *
   * @memberof PWABadge
   * @returns  {void} returns empty promise you can use for error handling
   */
  syncClearBadge() {
    this.__innerClear();
  }

  /**
   * Removes app's badge.
   *
   * @async
   * @memberof PWABadge
   * @returns  {Promise<void>} returns empty promise you can use for error handling
   */
  async asyncClearBadge() {
    return this.__innerClear();
  }
}

export default PWABadge;
